116.py

```diff
@@ -0,0 +1,35 @@
+"""
+# Definition for a Node.
+class Node:
+    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
+        self.val = val
+        self.left = left
+        self.right = right
+        self.next = next
+"""
+
+class Solution:
+    def connect(self, root: 'Node') -> 'Node':
+        if not root:
+            return root
+
+        queue = collections.deque([root])
+
+        while queue:
+            level_length = len(queue)
+
+            for i in range(level_length):
+                node = queue.popleft()
+
+                if i == level_length - 1:
```

Comment:
This `if-else` is a little awkward, and I am not a fan of accessing `queue[0]` in the loop body. 

Each level is essentially a linked list in this problem. Imagine if you already know your previous node when you are visiting the current node, how would you connect these two nodes together to make them a linked list? Then answer should be trivial. Now the problem becomes how to keep track of your previous node when traversing each level.